ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -348,6 +377,296 @@ protected Map<DataNodeId, List<RemoteReplicaInfo>> selectReplicas(Map<DataNodeId
     return replicas;
   }
 
+  /**
+   * Generates group ids for active iteration for replicas, all replicas are mapped to one group id only
+   * Generates group ids for all datanodes for standby no progress groups
+   * All group ids are unique
+   * All replicas are mapped to particular group id and each datanode has one stand by remote replica group id
+   * so we make less number of parallel calls to remote hosts
+   * @param groupIdToRemoteReplicaMap {@link Map} this map will be populated with group ids to replicas map
+   * @param standByNoProgressGroupId {@link Map}  this map will be populated with datanode to standby group id
+   */
+  void generateGroupIdsForReplicas(Map<Integer, List<RemoteReplicaInfo>> groupIdToRemoteReplicaMap,
+      Map<DataNodeId, Integer> standByNoProgressGroupId) {
+
+    Map<DataNodeId, List<RemoteReplicaInfo>> dataNodeToRemoteReplicaInfo = selectReplicas(getRemoteReplicaInfos());
+
+    int groupId = 0;
+
+    for (Map.Entry<DataNodeId, List<RemoteReplicaInfo>> entry : dataNodeToRemoteReplicaInfo.entrySet()) {
+      DataNodeId remoteHost = entry.getKey();
+      List<RemoteReplicaInfo> remoteReplicasPerNode = entry.getValue();
+
+      //for this data node break a larger array of remote replicas to smaller multiple arrays of predefined size given in config
+      List<List<RemoteReplicaInfo>> remoteReplicaSegregatedList =
+          maxReplicaCountPerRequest > 0 ? Utils.partitionList(remoteReplicasPerNode, maxReplicaCountPerRequest)
+              : Collections.singletonList(remoteReplicasPerNode);
```

Comment:
can we get rid of this config altogether ? and just put 1 replica in a group ?

Reply:
We need this so , at any point of time, we make less calls to other node.